### PR TITLE
Add unit tests for _paginate_with_dead_link_mask

### LINKED ISSUE #871: Unit tests for `_paginate_with_dead_link_mask`
## Description

Add unit tests for the `_paginate_with_dead_link_mask` function:

https://github.com/WordPress/openverse-api/blob/v2.5.10/api/catalog/api/controllers/search_controller.py#L40

## Additional context

Part of a series of unit test issues (see https://github.com/WordPress/openverse-api/issues/872 & https://github.com/WordPress/openverse-api/issues/873)

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Fixes #871 by @AetherUnbound

## Description

Wow, what a monster of a function. There are a lot of branches, each with their own possible edge cases. The third branch is by far the most complex and writing cases for it is likewise complex. I've had a hard time wrapping my head around possible edge cases, particularly because the third branch handles the edge cases so well, so each sub-branch is already exercising one possible edge case anyway. I think they're all sufficiently covered, but let me know if you notice any holes.

## Testing Instructions

Read the tests. They should pass CI and should make sense. Review the additional documentation I've added.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: api/test/unit/controllers/test_search_controller.py [hunk 1]
```diff
@@ -49,3 +56,313 @@ def test_get_result_and_page_count(total_hits, real_result_count, page_size, exp
...
+    number of defined cases is greater than the number of listed branch combinations.
+
+    Additionally, start_B + end_B is impossible. end_B can only happen when the query
+    mask contains _at least_ sufficient live bits to skip the previous pages and cover
+    the entirety of the presently requested page. This is checked by evaluating the sum
+    of the bits and only using the bits to find the end of the query if the mask covers
+    the full size of the query. If not, then the unmasked end is used. If start_B happens,
+    this means that the mask was not even sufficient to cover the previous pages + 1, just
+    the previous pages _exactly_. If you combine this information, it is clear that end_B
+    cannot happen when start_B happens becuase the conditions for start_B preclude the
```
Comment (obulat) [inline api/test/unit/controllers/test_search_controller.py:345]:
```suggestion
    cannot happen when start_B happens because the conditions for start_B preclude the
```

**Review by obulat — APPROVED** [review]
Thank you for providing such comprehensive documentation for this test! I could not think of any edge cases to add.

File: api/catalog/api/controllers/search_controller.py [hunk 2]
```diff
@@ -37,6 +37,20 @@ class RankFeature(Query):
...
+    from Elasticsearch under the following conditions:
+    1. There is no query mask
+    2. The lower index is beyond the scope of the existing query mask
+    3. The lower index is within the scope of the existing query mask
+    but the upper index exceeds it
+
+    In all these cases, the query mask is not used to calculate the upper index.
+    """
+    return ceil(page_size * page / (1 - DEAD_LINK_RATIO))
+
```
Comment (zackkrida) [inline api/catalog/api/controllers/search_controller.py:52]: ❤️

**Review by zackkrida — APPROVED** [review]
Excellent documentation, lgtm.
